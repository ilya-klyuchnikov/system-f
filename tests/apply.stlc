(let ((id (lam (x) x)(all (t) (t -> t))))
  id)

(let ((id (Lam (a) (lam ((x a)) x))))
  id)

(let ((apply
       (lam (fun arg) (fun arg))
       (all (a) (-> (-> a a) a a)))
      (id (lam (x) x)(all (t) (t -> t))))
  ; full "annotation"
  ((@ apply int) (@ id int) 1))


(let ((apply
       (lam (fun arg) (fun arg))
       (all (a) (-> (-> a a) a a)))
      (id (lam (x) x)(all (t) (t -> t))))
  ; partial "annotation/type application" - works
  (apply (@ id int) 1))

(let ((apply
       (lam (fun arg) (fun arg))
       (all (a) (-> (-> a a) a a)))
      (id (lam (x) x)(all (t) (t -> t))))
  ; another partial "annotation/type application" - doesn't work
  ((@ apply int) id 1))

(let ((apply
       (lam (fun arg) (fun arg))
       (all (a) (-> (-> a a) a a)))
      (id (lam (x) x)(all (t) (t -> t))))
  ; no "annotation" - doesn't work
  (apply id 1))

;; apply and containers
(let ((tapply
       (lam (fun arg) (fun arg))
       (all (a) (-> (-> (* a) (* a)) (* a) (* a))))
      (id (lam (x) x)(all (t) (t -> t))))
  ; partial "annotation/type application" - works
  (tapply (@ id (* int)) (tup 1)))

