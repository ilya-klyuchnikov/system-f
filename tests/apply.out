>  : (all (a) (-> a a))
-> #<function>
>  : (all (a) (-> a a))
-> #<function>
>  : int
-> 1
>  : int
-> 1
> type error: type mismatch: (-> int int) ≠ (all (a) (-> a a))
> type error: type mismatch: (-> (-> _ _) _ _) ≠ (-> (all (a) (-> a a)) int _)
>  : (* int)
-> (tup 1)
> 